rfcs/006-package-management.md

```diff
@@ -0,0 +1,243 @@
+---
+feature: package management
+start-date: 2024-06-28
+author: Joe Neeman
+---
+
+# Package management
+
+People want to reuse code, but Nickel doesn't currently have a good way to do
+it. We should have a way to fetch packages and make them easily available to
+Nickel code. The mechanism needs to be predictable (it should fetch the code
+that the user expects to fetch) and reliable (if it works on my machine it
+should work on your machine).
+
+## The manifest file
+
+We will require a manifest file in order to import packages. Manifest files must
+be named `package.ncl`, and they are found by searching up from the file being
+evaluated. That is, when the user invokes `nickel export path/to/foo.ncl`, we
+look for a manifest at `path/to/package.ncl` and then at `path/package.ncl`, and
+so on.
+
+The manifest file format is defined by the contract `std.package.Manifest`,
+while is defined as
+
+```nickel
+{
+  name | String,
+  version | Semver,
+  nickel-version | Semver,
+
+  dependencies
+    | { _:
+      [|
+        'Path String,
+        'Git { url | String, branch | optional | String, rev | optional | String },
+        'Index { name | String, version | Semver },
+      |]
+      }
+    | default
+    = {},
+}
+```
+
+So an example manifest might look like
+
+```nickel
+{
+  # ...
+  dependencies = {
+    foo = 'Index { package = "github/tweag/foo", version = "1.2.0" }
+    bar = 'Path "../my-bar",
+  }
+} | std.package.Manifest
+```
+
+### Alternative: inline dependencies
+
+Nix and Dhall allow for importing dependencies dynamically, using things like
+`fetchGit`. There was some discussion
+[here](https://github.com/tweag/nickel/issues/329#issuecomment-967372858)
+on the advantages and disadvantages of inline imports.
+
+### Alternative: toml manifest
+
+Maybe the manifest should be in some plain-data format like toml. This would
+be easier to modify programmatically.
+
+### Alternative: shorthand for registry imports
+
+Cargo allows a shorthand like
+
+```toml
+    "github/tweag/foo" = "1.2.0"
+```
+
+instead of
+
+```nickel
+    foo = 'Index { package = "github/tweag/foo", version = "1.2.0" }
+```
+
+Since we expect registry imports to be the common case, maybe it's worth having
+a shorthand?
+
+### Question: manifest file name
+
+The name `package.ncl` was chosen to be similar to npm's `package.json` or
+stack's `package.yaml`. One problem with this is that it could be confused for
+a nickel source file. Another possibility would be to use an extension-based
+name like `<package-name>.cabal`. Or just a stranger name that's less likely to
+conflict with something real.
+
+## Import statements
+
+The manifest file assigns a name to each dependency; to import the dependency
+named `foo` you simply write `import foo`. That is, an `import` statement either
+takes a string in quotes -- in which case it imports a path -- or an identifier
+without quote -- in which case it imports a package.
```

Comment:
I think there were discussions to make the import syntax more general, like (say we name our package manager `metalist`) `import foo from metalist`. Then, if we ever provide other integrations later on, you could do `import foo from flake` or something like that. What I like with the general import is that we're going the Nix route of names are resources that can be abstracted (or at least their resolution). Do you think that this will ever be useful? How does it fit in the current package management story?

Reply:
Ah, this is a good point. The syntax part is easy, but I think the main question is how integrated package management should be. I can see two main options:
1. package management is integrated into the interface (either built-in to the nickel cli, or by having a package-management cli that wraps the nickel cli)
2. package management is loosely coupled to nickel, where a package manager is expected to have some interface that nickel can use to query the map from package names to paths.

I think the choice mainly depends on how many different package managers we expect to have. If it's just `metallist` and `flake`, then maybe it makes sense to have them built-in. Last time we discussed this, I was under the impression that we could piggy-back off existing package repositories like `npm`. But since that seems not to be the case, I think there probably won't be many alternatives. (Which might be a good thing, as one of the concerns about the other plan was in fracturing the ecosystem)